primitives/src/merkle_tree/mod.rs

```diff
@@ -223,8 +227,23 @@ pub trait UniversalMerkleTreeScheme: MerkleTreeScheme {
     /// Update the leaf value at a given position
     /// * `pos` - zero-based index of the leaf in the tree
     /// * `elem` - newly updated element
-    fn update(&mut self, pos: Self::Index, elem: &Self::Element) -> Result<(), PrimitivesError>;
-
+    fn update(
+        &mut self,
+        pos: impl Borrow<Self::Index>,
+        elem: impl Borrow<Self::Element>,
+    ) -> LookupResult<Self::Element, ()>;
+
+    /// Build a universal merkle tree from a key-value set.
+    /// * `height` - height of the merkle tree
+    /// * `data` - an iterator of key-value pairs. Could be a std hashmap or
+    ///   simply an array or a slice of (key, value) pairs
+    fn from_kv_set<BI, BE>(
+        height: usize,
+        data: impl IntoIterator<Item = impl Borrow<(BI, BE)>>,
+    ) -> Result<Self, PrimitivesError>
+    where
+        BI: Borrow<Self::Index>,
+        BE: Borrow<Self::Element>;
```

Comment:
if `data` can already be turned into an iterator of tuple of references, 
we don't need the iterator item to be borrowed.

i.e. 
`data: impl IntoIterator<Item = (BI, BE)>` should be fine.

Reply:
It doesn't work for `&[(1,2)]`. In this case, you get references of tuples instead of tuple of references.
I was haunted by this before.